### PR TITLE
Detachable preview

### PR DESCRIPTION
**Pull request by K-LV**
WIP. Fixes humphd/brackets#189 and working off https://github.com/KeniMorri/brackets-browser-livedev/tree/k_experiments, see https://github.com/AJDBenner/brackets-browser-livedev/pull/7.

This patch allows for a detachable preview to be opened when a button is clicked. The detached preview window is in sync with the attached preview. The button acts as a toggle. So when it's clicked again, it closes the detached preview. Closing the detached window reopens the attached preview.  There is currently 1 bug, however. When you add/remove an element and close the detached window right away, the attached preview will not re-appear. Clicking the toggle button will show the attached preview. For the moment this patch only works in Chrome. There is an issue with FF and how the `beforeunload` event is triggered there.

### CODE REVIEW

**Comment by humphd** [comment]
Can we change the "reattach" button to `appbar.window.maximize` too?

**Comment by humphd** [comment]
Specifically, to this:

``` svg
<?xml version="1.0" encoding="utf-8"?>
<Canvas xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" x:Name="appbar_monitor" Width="76" Height="76" Clip="F1 M 0,0L 76,0L 76,76L 0,76L 0,0">
    <Path Width="39.9999" Height="33.9999" Canvas.Left="18" Canvas.Top="23.0002" Stretch="Fill" Fill="#FF000000" Data="F1 M 20,23.0002L 55.9998,23.0002C 57.1044,23.0002 57.9998,23.8956 57.9998,25.0002L 57.9999,46C 57.9999,47.1046 57.1045,48 55.9999,48L 41,48L 41,53L 45,53C 46.1046,53 47,53.8954 47,55L 47,57L 29,57L 29,55C 29,53.8954 29.8955,53 31,53L 35,53L 35,48L 20,48C 18.8954,48 18,47.1046 18,46L 18,25.0002C 18,23.8956 18.8954,23.0002 20,23.0002 Z M 21,26.0002L 21,45L 54.9999,45L 54.9998,26.0002L 21,26.0002 Z "/>
</Canvas>
```

File: lib/iframe-browser.js [hunk 1]
```diff
@@ -5,7 +5,9 @@ define(function (require, exports, module) {
 
     var CommandManager      = brackets.getModule("command/CommandManager"),
         MainViewManager     = brackets.getModule("view/MainViewManager"),
-        Commands            = brackets.getModule("command/Commands");
+        Commands            = brackets.getModule("command/Commands"),
+        Resizer     = brackets.getModule("utils/Resizer"),
```
Comment (humphd) [inline lib/iframe-browser.js]: Watch indents here, should match other code.

File: lib/iframe-browser.js [hunk 2]
```diff
@@ -128,6 +138,37 @@ define(function (require, exports, module) {
...
+    function attachPreview() {
+        if(detachedWindow && isReload) {
+            isReload = false;
+            return;
+        }
+
+        if(getDetachedPreview()) {
+            detachedWindow.removeEventListener("beforeunload", attachPreview, false);
+            detachedWindow.close();
+            detachedWindow = null;
```
Comment (humphd) [inline lib/iframe-browser.js]: Why do you do this?  I think that's wrong.  `getDetachedPreview()` will return nothing if it's `.closed===true`.

File: lib/iframe-browser.js [hunk 3]
```diff
@@ -128,6 +138,37 @@ define(function (require, exports, module) {
...
+        $("#liveDevButton").addClass("liveDevButtonDetach");
+        $("#first-pane").removeClass("expandEditor");
+
+        // Adds tooltip prompting user to detach preview
+        StatusBar.addIndicator("liveDevButtonBox", $("#liveDevButtonBox"), true, "",
+                               "Click to open preview in separate window", "status-overwrite"); 
+    }
+
+    function setListener() {
+        if(detachedWindow) {
```
Comment (humphd) [inline lib/iframe-browser.js]: I think probably using `getDetachedPreveiw()` everywhere is smart:

``` js
var detachedPreview = getDetachedPreview();
if(detachedPreview) {
...
}
```

File: lib/iframe-browser.js [hunk 4]
```diff
@@ -128,6 +138,37 @@ define(function (require, exports, module) {
...
+    /**
+     * Used to show second pane, change lilveDevButton background and close the detached preview
+     */
+    function attachPreview() {
+        if(detachedWindow && isReload) {
+            isReload = false;
+            return;
+        }
+
+        if(getDetachedPreview()) {
```
Comment (humphd) [inline lib/iframe-browser.js]: If you're going to use the method call here, you should cache the value:

``` js
var detachedPreview = getDetachedPreview();
if(detachedPreview) {
...
}
```

File: lib/iframe-browser.js [hunk 5]
```diff
@@ -128,6 +138,39 @@ define(function (require, exports, module) {
...
+        var detachedPreview = getDetachedPreview();
+        if(detachedPreview && isReload) {
+            isReload = false;
+            return;
+        }
+
+        if(detachedPreview) {
+            detachedWindow.removeEventListener("beforeunload", attachPreview, false);
+            detachedWindow.close();
+            // detachedWindow = null;
```
Comment (humphd) [inline lib/iframe-browser.js]: Remove this line if it's not needed anymore.

File: lib/iframe-browser.js [hunk 6]
```diff
@@ -88,9 +91,9 @@ define(function (require, exports, module) {
         if(iframe) {
             iframe.src = url;
         }
-
-        var detachedWindow = getDetachedPreview();
-        if(detachedWindow) {
+        var detachedPreview = getDetachedPreview();
+        if(detachedPreview) {
+            isReload = true;
             detachedWindow.location.replace(url);
```
Comment (humphd) [inline lib/iframe-browser.js]: Use `detachedPreview` here, since you just got it above.

File: lib/iframe-browser.js [hunk 7]
```diff
@@ -128,6 +138,39 @@ define(function (require, exports, module) {
...
+     */
+    function attachPreview() {
+        var detachedPreview = getDetachedPreview();
+        if(detachedPreview && isReload) {
+            isReload = false;
+            return;
+        }
+
+        if(detachedPreview) {
+            detachedWindow.removeEventListener("beforeunload", attachPreview, false);
```
Comment (humphd) [inline lib/iframe-browser.js]: Change this to `detachedPreview` and in the line below too.

File: lib/iframe-browser.js [hunk 8]
```diff
@@ -128,6 +138,39 @@ define(function (require, exports, module) {
...
+
+        // Adds tooltip prompting user to detach preview
+        StatusBar.addIndicator("liveDevButtonBox", $("#liveDevButtonBox"), true, "",
+                               "Click to open preview in separate window", "status-overwrite"); 
+    }
+
+    function setListener() {
+        var detachedPreview = getDetachedPreview();
+        if(detachedPreview) {
+            detachedWindow.addEventListener("beforeunload", attachPreview, false);
```
Comment (humphd) [inline lib/iframe-browser.js]: Same deal here.

**Comment by humphd** [comment]
I think this is r+.  Let's get a review from Gideon too.

**Comment by gideonthomas** [comment]
r+

**Pull request merged**